File: go/test/subscriber_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/statuspage-sdk/go"
	"github.com/voxgig-sdk/statuspage-sdk/go/core"

	vs "github.com/voxgig-sdk/statuspage-sdk/go/utility/struct"
)

func TestSubscriberEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Subscriber(nil)
		if ent == nil {
			t.Fatal("expected non-nil SubscriberEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"subscriber": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Subscriber(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Subscriber(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := subscriberBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "subscriber." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set STATUSPAGE_TEST_SUBSCRIBER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		subscriberRef01Ent := client.Subscriber(nil)
		subscriberRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "subscriber"}, setup.data), "subscriber_ref01"))
		subscriberRef01Data["incident_id"] = setup.idmap["incident01"]
		subscriberRef01Data["page_id"] = setup.idmap["page01"]

		subscriberRef01DataResult, err := subscriberRef01Ent.Create(subscriberRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		subscriberRef01Data = core.ToMapAny(subscriberRef01DataResult)
		if subscriberRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if subscriberRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		subscriberRef01Match := map[string]any{
			"page_id": setup.idmap["page01"],
		}

		subscriberRef01ListResult, err := subscriberRef01Ent.List(subscriberRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		subscriberRef01List, subscriberRef01ListOk := subscriberRef01ListResult.([]any)
		if !subscriberRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", subscriberRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(subscriberRef01List), map[string]any{"id": subscriberRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		subscriberRef01DataUp0Up := map[string]any{
			"id": subscriberRef01Data["id"],
			"page_id": setup.idmap["page_id"],
		}

		subscriberRef01MarkdefUp0Name := "component"
		subscriberRef01MarkdefUp0Value := fmt.Sprintf("Mark01-subscriber_ref01_%d", setup.now)
		subscriberRef01DataUp0Up[subscriberRef01MarkdefUp0Name] = subscriberRef01MarkdefUp0Value

		subscriberRef01ResdataUp0Result, err := subscriberRef01Ent.Update(subscriberRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		subscriberRef01ResdataUp0 := core.ToMapAny(subscriberRef01ResdataUp0Result)
		if subscriberRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if subscriberRef01ResdataUp0["id"] != subscriberRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if subscriberRef01ResdataUp0[subscriberRef01MarkdefUp0Name] != subscriberRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", subscriberRef01MarkdefUp0Name, subscriberRef01ResdataUp0[subscriberRef01MarkdefUp0Name])
		}

		// LOAD
		subscriberRef01MatchDt0 := map[string]any{
			"id": subscriberRef01Data["id"],
		}
		subscriberRef01DataDt0Loaded, err := subscriberRef01Ent.Load(subscriberRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		subscriberRef01DataDt0LoadResult := core.ToMapAny(subscriberRef01DataDt0Loaded)
		if subscriberRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if subscriberRef01DataDt0LoadResult["id"] != subscriberRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		subscriberRef01MatchRm0 := map[string]any{
			"id": subscriberRef01Data["id"],
		}
		_, err = subscriberRef01Ent.Remove(subscriberRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		subscriberRef01MatchRt0 := map[string]any{
			"page_id": setup.idmap["page01"],
		}

		subscriberRef01ListRt0Result, err := subscriberRef01Ent.List(subscriberRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		subscriberRef01ListRt0, subscriberRef01ListRt0Ok := subscriberRef01ListRt0Result.([]any)
		if !subscriberRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", subscriberRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(subscriberRef01ListRt0), map[string]any{"id": subscriberRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func subscriberBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "subscriber", "SubscriberTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read subscriber test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse subscriber test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"subscriber01", "subscriber02", "subscriber03", "page01", "page02", "page03", "incident01", "incident02", "incident03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("STATUSPAGE_TEST_SUBSCRIBER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"STATUSPAGE_TEST_SUBSCRIBER_ENTID": idmap,
		"STATUSPAGE_TEST_LIVE":      "FALSE",
		"STATUSPAGE_TEST_EXPLAIN":   "FALSE",
		"STATUSPAGE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["STATUSPAGE_TEST_SUBSCRIBER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add page_id alias for update test.
	if idmapResolved["page_id"] == nil {
		idmapResolved["page_id"] = idmapResolved["page01"]
	}

	if env["STATUSPAGE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["STATUSPAGE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewStatuspageSDK(core.ToMapAny(mergedOpts))
	}

	live := env["STATUSPAGE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["STATUSPAGE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/SubscriberEntityTest.php
<?php
declare(strict_types=1);

// Subscriber entity test

require_once __DIR__ . '/../statuspage_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class SubscriberEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = StatuspageSDK::test(null, null);
        $ent = $testsdk->Subscriber(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "subscriber" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = StatuspageSDK::test($seed, null);
        $seen = iterator_to_array($base->Subscriber(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = StatuspageConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = StatuspageSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Subscriber(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = subscriber_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "subscriber." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set STATUSPAGE_TEST_SUBSCRIBER_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $subscriber_ref01_ent = $client->Subscriber(null);
        $subscriber_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.subscriber"), "subscriber_ref01"));
        $subscriber_ref01_data["incident_id"] = $setup["idmap"]["incident01"];
        $subscriber_ref01_data["page_id"] = $setup["idmap"]["page01"];

        $subscriber_ref01_data_result = $subscriber_ref01_ent->create($subscriber_ref01_data, null);
        $subscriber_ref01_data = Helpers::to_map($subscriber_ref01_data_result);
        $this->assertNotNull($subscriber_ref01_data);
        $this->assertNotNull($subscriber_ref01_data["id"]);

        // LIST
        $subscriber_ref01_match = [
            "page_id" => $setup["idmap"]["page01"],
        ];

        $subscriber_ref01_list_result = $subscriber_ref01_ent->list($subscriber_ref01_match, null);
        $this->assertIsArray($subscriber_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($subscriber_ref01_list_result),
            ["id" => $subscriber_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $subscriber_ref01_data_up0_up = [
            "id" => $subscriber_ref01_data["id"],
            "page_id" => $setup["idmap"]["page_id"],
        ];

        $subscriber_ref01_markdef_up0_name = "component";
        $subscriber_ref01_markdef_up0_value = "Mark01-subscriber_ref01_" . $setup["now"];
        $subscriber_ref01_data_up0_up[$subscriber_ref01_markdef_up0_name] = $subscriber_ref01_markdef_up0_value;

        $subscriber_ref01_resdata_up0_result = $subscriber_ref01_ent->update($subscriber_ref01_data_up0_up, null);
        $subscriber_ref01_resdata_up0 = Helpers::to_map($subscriber_ref01_resdata_up0_result);
        $this->assertNotNull($subscriber_ref01_resdata_up0);
        $this->assertEquals($subscriber_ref01_resdata_up0["id"], $subscriber_ref01_data_up0_up["id"]);
        $this->assertEquals($subscriber_ref01_resdata_up0[$subscriber_ref01_markdef_up0_name], $subscriber_ref01_markdef_up0_value);

        // LOAD
        $subscriber_ref01_match_dt0 = [
            "id" => $subscriber_ref01_data["id"],
        ];
        $subscriber_ref01_data_dt0_loaded = $subscriber_ref01_ent->load($subscriber_ref01_match_dt0, null);
        $subscriber_ref01_data_dt0_load_result = Helpers::to_map($subscriber_ref01_data_dt0_loaded);
        $this->assertNotNull($subscriber_ref01_data_dt0_load_result);
        $this->assertEquals($subscriber_ref01_data_dt0_load_result["id"], $subscriber_ref01_data["id"]);

        // REMOVE
        $subscriber_ref01_match_rm0 = [
            "id" => $subscriber_ref01_data["id"],
        ];
        $subscriber_ref01_ent->remove($subscriber_ref01_match_rm0, null);

        // LIST
        $subscriber_ref01_match_rt0 = [
            "page_id" => $setup["idmap"]["page01"],
        ];

        $subscriber_ref01_list_rt0_result = $subscriber_ref01_ent->list($subscriber_ref01_match_rt0, null);
        $this->assertIsArray($subscriber_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($subscriber_ref01_list_rt0_result),
            ["id" => $subscriber_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function subscriber_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/subscriber/SubscriberTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = StatuspageSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["subscriber01", "subscriber02", "subscriber03", "page01", "page02", "page03", "incident01", "incident02", "incident03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("STATUSPAGE_TEST_SUBSCRIBER_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "STATUSPAGE_TEST_SUBSCRIBER_ENTID" => $idmap,
        "STATUSPAGE_TEST_LIVE" => "FALSE",
        "STATUSPAGE_TEST_EXPLAIN" => "FALSE",
        "STATUSPAGE_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["STATUSPAGE_TEST_SUBSCRIBER_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["page_id"])) {
        $idmap_resolved["page_id"] = $idmap_resolved["page01"];
    }

    if ($env["STATUSPAGE_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["STATUSPAGE_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new StatuspageSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["STATUSPAGE_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["STATUSPAGE_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
